File: src/main/scala/uk/co/ridentbyte/model/Http.scala
package uk.co.ridentbyte.model

import java.net.{MalformedURLException, URI, URL, URLDecoder}

import org.apache.http.client.methods._
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.{HttpClientBuilder, LaxRedirectStrategy}
import org.apache.http.util.EntityUtils

import scala.collection.JavaConverters._

case class Http(request: CardinalRequest) {

  private def parseURI(rawUri: String): String = {
    val rawUriWithProtocol = try {
      new URL(rawUri)
      new URI(rawUri)
      rawUri
    } catch {
      case _: MalformedURLException => "http://" + rawUri
    }
    val decodedURL = URLDecoder.decode(rawUriWithProtocol, "UTF-8")
    val url = new URL(decodedURL)
    val uri = new URI(url.getProtocol, url.getUserInfo, url.getHost, url.getPort, url.getPath, url.getQuery, url.getRef)
    uri.toASCIIString
  }

  def send: CardinalHttpResponse = {
    val parsedUri = parseURI(request.uri)
    val splitHeaders = request.headers.map { header =>
      val splitHeader = header.split(":")
      if (splitHeader.length == 2) {
        (splitHeader(0), splitHeader(1))
      } else {
        (splitHeader(0), "")
      }
    }
    
    request.verb match {
      case "POST" => {

        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpPost = new HttpPost(parsedUri)
        splitHeaders.foreach { header =>
          httpPost.setHeader(header._1, header._2)
        }
        if (request.body.isDefined) {
          httpPost.setEntity(new StringEntity(request.body.get))
        }

        val res = httpClient.execute(httpPost)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "PUT" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpPut = new HttpPut(parsedUri)
        splitHeaders.foreach { header =>
          httpPut.setHeader(header._1, header._2)
        }

        if (request.body.isDefined) {
          httpPut.setEntity(new StringEntity(request.body.get))
        }

        val res = httpClient.execute(httpPut)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "GET" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpGet = new HttpGet(parsedUri)
        splitHeaders.foreach { header =>
          httpGet.setHeader(header._1, header._2)
        }

        val res = httpClient.execute(httpGet)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "DELETE" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpDelete = new HttpDelete(parsedUri)
        splitHeaders.foreach { header =>
          httpDelete.setHeader(header._1, header._2)
        }

        val res = httpClient.execute(httpDelete)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "HEAD" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpHead = new HttpHead(parsedUri)
        splitHeaders.foreach { header =>
          httpHead.setHeader(header._1, header._2)
        }

        val res = httpClient.execute(httpHead)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "OPTIONS" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpOptions = new HttpOptions(parsedUri)
        splitHeaders.foreach { header =>
          httpOptions.setHeader(header._1, header._2)
        }

        val res = httpClient.execute(httpOptions)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "TRACE" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpTrace = new HttpTrace(parsedUri)
        splitHeaders.foreach { header =>
          httpTrace.setHeader(header._1, header._2)
        }

        val res = httpClient.execute(httpTrace)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

      case "PATCH" => {
        val httpClient = HttpClientBuilder.create().setRedirectStrategy(new LaxRedirectStrategy()).build()

        val httpPatch = new HttpPatch(parsedUri)
        splitHeaders.foreach { header =>
          httpPatch.setHeader(header._1, header._2)
        }

        if (request.body.isDefined) {
          httpPatch.setEntity(new StringEntity(request.body.get))
        }

        val res = httpClient.execute(httpPatch)
        val headers = res.getAllHeaders.map(h => (h.getName, h.getValue)).toMap.asJava
        val body = if (res.getEntity != null ) EntityUtils.toString(res.getEntity) else ""
        val code = res.getStatusLine.getStatusCode

        new CardinalHttpResponse(body, headers, code)
      }

    }

  }

}

File: src/main/java/uk/co/ridentbyte/model/Http.java
package uk.co.ridentbyte.model;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLDecoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.AbstractMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Http {

    private CardinalRequest request;

    public Http(CardinalRequest request) {
        this.request = request;
    }

    public CardinalHttpResponse send() {
        String parsedUri = parseUri(request.getUri());
        Map<String, String> splitHeaders = request.getHeaders().stream().map((h) -> {
           String[] splitHeader = h.split(":");
           if (splitHeader.length == 2) {
               return new AbstractMap.SimpleEntry<>(splitHeader[0], splitHeader[1]);
           } else {
               return new AbstractMap.SimpleEntry<>(splitHeader[0], "");
           }
        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));

        try {
            HttpClient httpClient = null;
            if (request.shouldFollowRedirects()) {
                httpClient = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.ALWAYS).build();
            } else {
                httpClient = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).build();
            }
            switch(request.getVerb()) {
                case "POST": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.POST(HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder = httpRequestBuilder.POST(HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "PUT": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.PUT(HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder = httpRequestBuilder.PUT(HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "GET": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri)).GET();

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "DELETE": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri)).DELETE();

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "HEAD": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.method("HEAD", HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder.method("HEAD", HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "OPTIONS": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.method("OPTIONS", HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder.method("OPTIONS", HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "TRACE": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.method("TRACE", HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder.method("TRACE", HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                case "PATCH": {
                    var httpRequestBuilder = HttpRequest.newBuilder().uri(URI.create(parsedUri));
                    if (request.getBody() != null) {
                        httpRequestBuilder = httpRequestBuilder.method("PATCH", HttpRequest.BodyPublishers.ofString(request.getBody()));
                    } else {
                        httpRequestBuilder.method("PATCH", HttpRequest.BodyPublishers.noBody());
                    }

                    for (Map.Entry<String, String> header : splitHeaders.entrySet()) {
                        httpRequestBuilder = httpRequestBuilder.header(header.getKey(), header.getValue());
                    }

                    HttpResponse<String> response = httpClient.send(httpRequestBuilder.build(), HttpResponse.BodyHandlers.ofString());
                    var responseHeaders = new HashMap<String, String>();
                    for (Map.Entry<String, List<String>> header : response.headers().map().entrySet()) {
                        responseHeaders.put(header.getKey(), String.join("", header.getValue()));
                    }
                    return new CardinalHttpResponse(response.body(), responseHeaders, response.statusCode());
                }

                default:
                    throw new RuntimeException("Invalid HTTP method");
            }
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }

    }

    private String parseUri(String rawUri) {
        String validUri = "";
        try {
            new URL(rawUri);
            validUri = rawUri;
        } catch (MalformedURLException e) {
            validUri =  "http://" + rawUri;
        }

        try {
            String decodedUri = URLDecoder.decode(validUri, "UTF-8");
            URL url = new URL(decodedUri);
            URI uri = new URI(
                    url.getProtocol(),
                    url.getUserInfo(),
                    url.getHost(),
                    url.getPort(),
                    url.getPath(),
                    url.getQuery(),
                    url.getRef()
            );
            return uri.toASCIIString();
        } catch (UnsupportedEncodingException | MalformedURLException | URISyntaxException e) {
            e.printStackTrace();
            return "";
        }
    }

}
